torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -0,0 +1,92 @@
+
+#include "torch_xla/csrc/xla_sharding_util.h"
+
+#include <unordered_map>
+
+#include "tensorflow/compiler/xla/execution_options_util.h"
+#include "tensorflow/compiler/xla/service/hlo_module.h"
+#include "tensorflow/compiler/xla/service/hlo_parser.h"
+#include "tensorflow/compiler/xla/service/hlo_pass_pipeline.h"
+#include "tensorflow/compiler/xla/service/hlo_verifier.h"
+#include "tensorflow/compiler/xla/service/sharding_propagation.h"
+#include "tensorflow/compiler/xla/service/spmd/spmd_partitioner.h"
+#include "tensorflow/compiler/xla/xla.pb.h"
+
+namespace torch_xla {
+namespace {
+
+using xla::internal::XlaBuilderFriend;
+
+}  // namespace
+
+void ShardingUtil::SetHloSharding(LoweringContext* lowering_ctx) {
+  for (std::pair<torch::lazy::Output, xla::XlaOp> elem :
+       lowering_ctx->GetEmittedOutputs()) {
+    const torch::lazy::Node* node = elem.first.node;
+    const XlaNode* xla_node = dynamic_cast<const XlaNode*>(node);
+    auto instruction = XlaBuilderFriend::GetInstruction(elem.second);
+    if (xla_node->GetSharding() != nullptr) {
+      //*instruction->mutable_sharding() = *xla_node->GetSharding();
+      {
+        // Annotate the full-shape input with the sharding.
+        xla::XlaScopedShardingAssignment assign_sharding(
+            lowering_ctx->builder(), *xla_node->GetSharding());
+        xla::CustomCall(lowering_ctx->builder(),
+                        /*call_target_name=*/"Sharding",
+                        xla_node->Lower(lowering_ctx), xla_node->xla_shape(),
+                        /*opaque=*/"");
+      }
+    }
+  }
+}
+
+xla::HloModuleProto ShardingUtil::SpmdPartitioningPass(
```

Comment:
This is a bit weird, I expect us to set a flag to let XLA compiler modify the HLO to be SPMD ready. Do we need to manually add the SpmdPartitioningPass here?

Reply:
Yes for xrt, since we don't have a dedicated spmd mode. This is also a valuable utility function for us to generate and debug partitioned graphes.